From: Anshuman Khandual <anshuman.khandual@arm.com>
Subject: mm/hugetlb: ensure adequate CMA areas available for hugetlb_cma[]
Date: Thu, 15 Feb 2024 10:04:05 +0530

HugeTLB CMA area array is being created for possible MAX_NUMNODES without
ensuring corresponding MAX_CMA_AREAS support in CMA.  This fails the build
for such scenarios indicating need for CONFIG_CMA_AREAS adjustment.

Link: https://lkml.kernel.org/r/20240215043405.2379295-1-anshuman.khandual@arm.com
Signed-off-by: Anshuman Khandual <anshuman.khandual@arm.com>
Cc: Muchun Song <muchun.song@linux.dev>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/hugetlb.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/mm/hugetlb.c~mm-hugetlb-ensure-adequate-cma-areas-available-for-hugetlb_cma
+++ a/mm/hugetlb.c
@@ -7855,6 +7855,13 @@ void __init hugetlb_cma_reserve(int orde
 	}
 
 	reserved = 0;
+
+	/*
+	 * There needs to be enough MAX_CMA_AREAS to accommodate
+	 * MAX_NUMNODES heap areas being created here. Otherwise
+	 * adjust CONFIG_CMA_AREAS as required.
+	 */
+	BUILD_BUG_ON(MAX_CMA_AREAS < MAX_NUMNODES);
 	for_each_online_node(nid) {
 		int res;
 		char name[CMA_MAX_NAME];
_
